internal/injector/injector.go

```diff
@@ -9,294 +9,193 @@
-		for path, name := range builtin.RestorerMap {
-			restorerMap[path] = name
-		}
-
-		for _, pkg := range pkgs {
-			if len(pkg.Errors) > 0 {
-				errs := make([]error, len(pkg.Errors))
-				for i := range pkg.Errors {
-					errs[i] = pkg.Errors[i]
-				}
-				return nil, errors.Join(errs...)
-			}
+	}
 
-			if pkgPath == "" {
-				// The first non-blank package path is the "top level" one (the one we care about).
-				pkgPath = pkg.PkgPath
-			}
+	fset := token.NewFileSet()
+	astFiles, err := parseFiles(fset, files)
+	if err != nil {
+		return nil, err
+	}
 
-			for _, imp := range pkg.Imports {
-				if imp.Name == "" {
-					// Happens when there is an error while processing the import, typically inability to resolve the name due to
-					// a typo or something. If we allow blank names in the map, the restorer just removes the qualifiers, which is
-					// obviously undesirable.
-					continue
-				}
-				restorerMap[imp.PkgPath] = imp.Name
-			}
+	uses, err := i.typeCheck(fset, astFiles)
+	if err != nil {
+		return nil, err
+	}
 
-			// pkg.Decorator is nil in case the package in question does not include any go source file. This can be the case
-			// when building test packages that do not include any non-test source file; in which case the "package under
-			// test" is empty. This is because the loader returns three different entries when processing tests:
-			// 1. The package under test (which may be empty)
-			// 2. The test functions
-			// 3. The test binary/main
-			if pkg.Decorator != nil {
-				decorators = append(decorators, pkg.Decorator)
-			}
+	decorator := decorator.NewDecoratorWithImports(fset, i.ImportPath, gotypes.New(uses))
+	dstFiles := make([]*dst.File, len(astFiles))
+	for idx, astFile := range astFiles {
+		dstFiles[idx], err = decorator.DecorateFile(astFile)
+		if err != nil {
+			return nil, err
 		}
 	}
 
-	if len(decorators) == 0 {
-		return nil, errors.New("no decorators could be created")
+	result := make(map[string]InjectedFile, len(files))
+	for idx, dstFile := range dstFiles {
+		res, err := i.injectFile(decorator, dstFile)
+		if err != nil {
+			return nil, err
+		}
+		if res.Modified {
+			result[files[idx]] = res.InjectedFile
+		}
 	}
-
-	return &Injector{
-		fileset:    fileset,
-		decorators: decorators,
-		newRestorer: func(filename string) *decorator.FileRestorer {
-			res := decorator.NewRestorerWithImports(pkgPath, guess.WithMap(restorerMap)).FileRestorer()
-			res.Name = filename
-			return res
-		},
-		opts: opts,
-	}, nil
+	return result, nil
 }
 
-type (
-	// Result describes the result of an injection operation.
-	Result struct {
-		References typed.ReferenceMap // New package references injected into the file and what kind of reference they are
-		Filename   string             // The file name of the output file (may be different from the input file)
-		Modified   bool               // Whether the file was modified
+func (i *Injector) validate() error {
```

Comment:
I've intentionally gone for something lighter weight than a full on functional options set-up, because this is more similar to what `packages.Load` does (and other stuff in the compiler-adjacent ecosystem); but I decided to have a `validate` function to produce proper error messaging instead of doing what the stuff in the other similar packages here do (that is, silently not do what you meant, and leave you scratching your head as to why)